distributed/worker.py

```diff
@@ -1199,22 +1266,19 @@ async def heartbeat(self):
         finally:
             self.heartbeat_active = False
 
+    @fail_hard
     async def handle_scheduler(self, comm):
-        try:
-            await self.handle_stream(comm, every_cycle=[self.ensure_communicating])
-        except Exception as e:
-            logger.exception(e)
-            raise
-        finally:
-            if self.reconnect and self.status in Status.ANY_RUNNING:
-                logger.info("Connection to scheduler broken.  Reconnecting...")
-                self.loop.add_callback(self.heartbeat)
-            else:
-                logger.info(
-                    "Connection to scheduler broken. Closing without reporting.  Status: %s",
-                    self.status,
-                )
-                await self.close(report=False)
+        await self.handle_stream(comm, every_cycle=[self.ensure_communicating])
+
+        if self.reconnect and self.status in Status.ANY_RUNNING:
+            logger.info("Connection to scheduler broken.  Reconnecting...")
+            self.loop.add_callback(self.heartbeat)
+        else:
+            logger.info(
+                "Connection to scheduler broken. Closing without reporting.  Status: %s",
+                self.status,
+            )
+            await self.close(report=False)
```

Comment:
You are now relying on this pass https://github.com/dask/distributed/blob/f15158324e9bf9abc9ae8de0f2a1cb59582de8ba/distributed/core.py#L636-L638

I think we should replace the FIXME

Reply:
Replace with what?  My read right now is that if the stream closes for some reason that looks like of like a comm error, we choose to let this close peacefully and then expect things to probably restart. 

I don't necessarily agree or disagree with this choice.  My gut says to leave it be for this PR.

However, it could be that some change here has magnified the importance of this choice.  Is that what you're suggesting?  If so, can you help me understand?  Or do you have a recommendation?

Reply:
Removed comment